### PR TITLE
Initial set of additions and fixes from Rivos for review

### PR DESCRIPTION
**Pull request by jerinjoy**
We want to restart this conversation about working on a single RVV implementation to be merged into upstream Gem5.

We also think adding microcode support for the RVV instructions is the right way to go.

These are an initial set of changes from Rivos for review. We plan to add more instructions going forward.

### CODE REVIEW

**Comment by ksco** [comment]
Thanks for the contribution, overall looks good.

File: configs/common/Simulation.py [hunk 1]
```diff
@@ -733,4 +733,4 @@ def run(options, root, testsys, cpu_class):
         m5.checkpoint(joinpath(cptdir, "cpt.%d"))
 
     if exit_event.getCode() != 0:
-        print("Simulated exit code not 0! Exit code is", exit_event.getCode())
+        fatal("Simulated exit code is %s" % exit_event.getCode())
```
Comment (ksco) [inline configs/common/Simulation.py]: This looks like an unrelated change.

File: src/arch/riscv/regs/vector.hh [hunk 2]
```diff
@@ -47,10 +48,12 @@ namespace RiscvISA
 
 constexpr unsigned NumVecElemPerVecReg = 4;
 using VecElem = uint64_t;
-constexpr size_t vlenb = NumVecElemPerVecReg * sizeof(VecElem);
-constexpr size_t VLEN = vlenb * 8;
+constexpr size_t VLENB = NumVecElemPerVecReg * sizeof(VecElem);
```
Comment (ksco) [inline src/arch/riscv/regs/vector.hh:51]: `vlenb` is intentionally lowercase in the spec, I think it's because the register names are lowercase. Maybe we should keep it lowercase.

File: configs/common/Simulation.py (hunk 1 above)
Comment (jerinjoy) [reply configs/common/Simulation.py]: Yes I was using this to detect when a test failed by checking the process return code. I've dropped this commit.

File: src/arch/riscv/regs/vector.hh (hunk 2 above)
Comment (jerinjoy) [reply src/arch/riscv/regs/vector.hh:51]: I thought it would be clearer that this was a global variable when it was referenced in other files. I'm not sure if Gem5 has a convention regarding this. All the other globals appeared to be upper case.

Example:

```
    for (size_t i = 0; i < VLENB; i++) {
        %(memacc_code)s;
    }
```

I realize I also changed this string be uppercase, which I shouldn't have:

```
    {CSR_VTYPE,        {"vtype" , MISCREG_VTYPE}},
    {CSR_VLENB,        {"VLENB" , MISCREG_VLENB}}
```

Should I change this back to lower case and keep the rest of the uppercase changes or should I change everything back to lowercase?

File: src/arch/riscv/insts/vector.cc [hunk 3]
```diff
@@ -41,6 +41,74 @@ namespace gem5
 namespace RiscvISA
 {
 
+void roundUnsignedInteger(__uint128_t &result, uint32_t xrm, int gb) {
```
Comment (ksco) [inline src/arch/riscv/insts/vector.cc:44]: roundUnsignedInteger & roundSignedInteger looks exactly the same. And for utility functions, it might be more appropriate to put it in riscv/utility.hh.

File: src/arch/riscv/insts/vector.cc [hunk 4]
```diff
@@ -41,6 +41,74 @@ namespace gem5
...
+      case VectorRoundingMode::RoundToOdd:
+        if (result & (lsb - 1)) {
+          result |= lsb;
+        }
+        break;
+      default:
+        printf("error: unknown vector rounding mode\n");
+        exit(1);
+    }
+  } while (0);
```
Comment (ksco) [inline src/arch/riscv/insts/vector.cc]: We don't need `do while` here since it's not a macro.

File: src/arch/riscv/regs/vector.hh (hunk 2 above)
Comment (ksco) [reply src/arch/riscv/regs/vector.hh:51]: Yes, I think change the register name back to lowercase and leave other parts as they are should be fine.

Comment (jerinjoy) [reply src/arch/riscv/regs/vector.hh:51]: For now I've just changed this back to lowercase:

```
    {CSR_VTYPE,        {"vtype" , MISCREG_VTYPE}},
    {CSR_VLENB,        {"vlenb" , MISCREG_VLENB}}
```

File: src/arch/riscv/insts/vector.cc (hunk 4 above)
Comment (jerinjoy) [reply src/arch/riscv/insts/vector.cc]: Fixed. I took this code from Spike where they are macros.

File: src/arch/riscv/insts/vector.cc (hunk 3 above)
Comment (ksco) [reply src/arch/riscv/insts/vector.cc:44]: We seem to have implemented this function as well (haven't synced to GitHub yet) and may need to do some rebase after this PR is merged, but here's our implementation.

```cpp
template<typename T> T
int_rounding(T result, uint8_t xrm, unsigned gb) {
    const uint64_t lsb = 1UL << gb;
    const uint64_t lsb_half = lsb >> 1;
    switch (xrm) {
    case 0 /* RNU */:
        result += lsb_half;
        break;
    case 1 /* RNE */:
        if ((result & lsb_half) &&
            ((result & (lsb_half - 1)) || (result & lsb)))
            result += lsb;
        break;
    case 2 /* RDN */:
        break;
    case 3 /* ROD */:
        if (result & (lsb - 1))
            result |= lsb;
        break;
    default:
        panic("Invalid xrm value %d", (int)xrm);
    }

    return result;
}
```

Comment (ksco) [reply src/arch/riscv/insts/vector.cc:44]: But it's ok to keep it unchanged though, we can't avoid modifying it in a later rebase anyway.

**Comment by ksco** [comment]
Hi @jerinjoy, I've rebased PLCT's unsynced work to GitHub. Now we're on the same track, yay!

cc @huxuan0307 @XDUFanYang @joy2myself

**Comment by ksco** [comment]
@jerinjoy Also, do you mind doing some code reviews for future PRs?

**Comment by jerinjoy** [comment]
@ksco thanks for rebasing! I'll pull the changes and make future changes on top of these.

I'm happy to do the code reviews.
